### PR TITLE
Additional boundary-layer experiments 

### PR DESCRIPTION
**Pull request by akshaysridhar**
1) Stable BL (Kosovic and Curry, 2000)
2) Unstable BL (Kitamura and Ito, 2016)
	new file:   experiments/AtmosLES/stable_bl_kitamura.jl
	new file:   experiments/AtmosLES/unstable_bl_kitamura.jl

# Description

This PR adds two experiments, with emphasis on surface fluxes and surface layer dynamics. 

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: experiments/AtmosLES/stable_bl_kosovic.jl [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+    # Prescribe domain parameters
+    xmax = FT(100)
+    ymax = FT(100)
+    zmax = FT(400)
+
+    t0 = FT(0)
+
+    # For a full-run, please set the timeend to 3600*6 seconds
+    # For the test we set this to == 30 minutes
```
Comment (szy21) [inline experiments/AtmosLES/stable_bl_kosovic.jl:336]: It seems you copied the comments from bomex, which is not consistent with the code :)

**Comment by kpamnany** [comment]
After #1415, `init_state_prognostic` and `boundarycondition` must now be in a subtype of `AbstractAtmosProblem`. That must be passed to `AtmosModel` as the `problem` parameter (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L302-L326) for example). Also, `init_state_prognostic` must now have an additional first parameter -- `problem` (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L196) for example).

You can use the available `AtmosProblem` subtype of `AbstractAtmosProblem` for ease, but the point of this abstraction is to let you define your own problem types and share them between experiments. Some examples of this (for the GCM and for BOMEX) are incoming.

**Comment by akshaysridhar** [comment]
> After #1415, `init_state_prognostic` and `boundarycondition` must now be in a subtype of `AbstractAtmosProblem`. That must be passed to `AtmosModel` as the `problem` parameter (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L302-L326) for example). Also, `init_state_prognostic` must now have an additional first parameter -- `problem` (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L196) for example).
> 
> You can use the available `AtmosProblem` subtype of `AbstractAtmosProblem` for ease, but the point of this abstraction is to let you define your own problem types and share them between experiments. Some examples of this (for the GCM and for BOMEX) are incoming.



> After #1415, `init_state_prognostic` and `boundarycondition` must now be in a subtype of `AbstractAtmosProblem`. That must be passed to `AtmosModel` as the `problem` parameter (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L302-L326) for example). Also, `init_state_prognostic` must now have an additional first parameter -- `problem` (see in [`dycoms.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/AtmosLES/dycoms.jl#L196) for example).
> 
> You can use the available `AtmosProblem` subtype of `AbstractAtmosProblem` for ease, but the point of this abstraction is to let you define your own problem types and share them between experiments. Some examples of this (for the GCM and for BOMEX) are incoming.

Noted, thanks!

File: experiments/AtmosLES/stable_bl_kosovic.jl [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+                momentum = Impenetrable(DragLaw(
+                    # normPu_int is the internal horizontal speed
+                    # P represents the projection onto the horizontal
+                    (state, aux, t, normPu_int) -> (u_star / normPu_int)^2,
+                )),
+                energy = BulkFormulaEnergy(
+                    (state, aux, t, normPu_int) -> C_drag,
+                    (state, aux, t) ->
+                        surface_temperature_variation(state, aux, t),
+                    (state, aux, t) -> q_sfc,
```
Comment (kpamnany) [inline experiments/AtmosLES/stable_bl_kosovic.jl:272]:
```suggestion
                    (state, aux, t) ->
                        (surface_temperature_variation(state, aux, t), q_sfc),
```
Due to #1454.

File: experiments/AtmosLES/unstable_bl_kitamura.jl [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+                momentum = Impenetrable(DragLaw(
+                    # normPu_int is the internal horizontal speed
+                    # P represents the projection onto the horizontal
+                    (state, aux, t, normPu_int) -> (u_star / normPu_int)^2,
+                )),
+                energy = BulkFormulaEnergy(
+                    (state, aux, t, normPu_int) -> C_drag,
+                    (state, aux, t) ->
+                        surface_temperature_variation(state, aux, t),
+                    (state, aux, t) -> q_sfc,
```
Comment (kpamnany) [inline experiments/AtmosLES/unstable_bl_kitamura.jl:269]:
```suggestion
                    (state, aux, t) ->
                        (surface_temperature_variation(state, aux, t), q_sfc),
```
Due to #1454.

File: experiments/AtmosLES/stable_bl_kosovic.jl [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+                ),
+                moisture = BulkFormulaMoisture(
+                    (state, aux, t, normPu_int) -> C_drag,
+                    (state, aux, t) -> q_sfc,
```
Comment (kpamnany) [inline experiments/AtmosLES/stable_bl_kosovic.jl:276]: @szy21: if we had changed `bc_moisture.jl` also in #1454, then we would have needed to call `surface_temperature_variation()` even though the temperature is not required in the moisture flux computation, right?

File: experiments/AtmosLES/unstable_bl_kitamura.jl [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+using LinearAlgebra
+
+using CLIMAParameters
+using CLIMAParameters.Planet: R_d, cp_d, cv_d, MSLP, grav, day
+struct EarthParameterSet <: AbstractEarthParameterSet end
+const param_set = EarthParameterSet()
+
+import ClimateMachine.BalanceLaws: vars_state_prognostic
+import ClimateMachine.Atmos: source!, atmos_source!, altitude
+import ClimateMachine.Atmos: flux_second_order!, thermo_state
```
Comment (charleskawczynski) [inline experiments/AtmosLES/unstable_bl_kitamura.jl:62]:
```suggestion
import ClimateMachine.Atmos: atmos_source!, flux_second_order!
using ClimateMachine.Atmos: altitude, thermo_state
```

File: experiments/AtmosLES/stable_bl_kosovic.jl [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+using LinearAlgebra
+
+using CLIMAParameters
+using CLIMAParameters.Planet: R_d, cp_d, cv_d, MSLP, grav, day
+struct EarthParameterSet <: AbstractEarthParameterSet end
+const param_set = EarthParameterSet()
+
+import ClimateMachine.BalanceLaws: vars_state_prognostic
+import ClimateMachine.Atmos: source!, atmos_source!, altitude
+import ClimateMachine.Atmos: flux_second_order!, thermo_state
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_kosovic.jl:61]:
```suggestion
import ClimateMachine.Atmos: atmos_source!, flux_second_order!
using ClimateMachine.Atmos: altitude, thermo_state
```

Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_kosovic.jl:61]:
```suggestion
import ClimateMachine.BalanceLaws: vars_state
using ClimateMachine.BalanceLaws: Prognostic
import ClimateMachine.Atmos: atmos_source!, flux_second_order!
using ClimateMachine.Atmos: altitude, thermo_state
```

File: experiments/AtmosLES/unstable_bl_kitamura.jl (hunk 5 above)
Comment (charleskawczynski) [inline experiments/AtmosLES/unstable_bl_kitamura.jl:62]:
```suggestion
import ClimateMachine.BalanceLaws: vars_state
using ClimateMachine.BalanceLaws: Prognostic
import ClimateMachine.Atmos: atmos_source!, flux_second_order!
using ClimateMachine.Atmos: altitude, thermo_state
```

File: experiments/AtmosLES/unstable_bl_kitamura.jl [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+function surface_temperature_variation(state, aux, t)
+    FT = eltype(state)
+    ρ = state.ρ
+    q_tot = state.moisture.ρq_tot / ρ
+    θ_liq_sfc = FT(291.15) + FT(20) * sinpi(FT(t / 12 / 3600))
+    TS = LiquidIcePotTempSHumEquil(param_set, θ_liq_sfc, ρ, q_tot)
+    return air_temperature(TS)
+end
+
+function config_problem(FT, N, resolution, xmax, ymax, zmax)
```
Comment (charleskawczynski) [inline experiments/AtmosLES/unstable_bl_kitamura.jl:211]:
```suggestion
function config_problem(::Type{FT}, N, resolution, xmax, ymax, zmax) where {FT}
```

File: experiments/AtmosLES/stable_bl_kosovic.jl [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+function surface_temperature_variation(state, aux, t)
+    FT = eltype(state)
+    ρ = state.ρ
+    q_tot = state.moisture.ρq_tot / ρ
+    θ_liq_sfc = FT(265) - FT(1 / 4) * (t / 3600)
+    TS = LiquidIcePotTempSHumEquil(param_set, θ_liq_sfc, ρ, q_tot)
+    return air_temperature(TS)
+end
+
+function config_problem(FT, N, resolution, xmax, ymax, zmax)
```
Comment (charleskawczynski) [inline experiments/AtmosLES/stable_bl_kosovic.jl:214]:
```suggestion
function config_problem(::Type{FT}, N, resolution, xmax, ymax, zmax) where {FT}
```

**Comment by kpamnany** [comment]
@akshaysridhar: rebased, updated to use `AtmosProblem` and squashed. Ran each experiment for about 2 minutes of wallclock time; they seem to run fine.

File: docs/list_of_apis.jl [hunk 9]
```diff
@@ -16,12 +16,9 @@ apis = [
         "Land Model" => "APIs/Land/LandModel.md",
         "Soil Water Parameterizations" =>
             "APIs/Land/SoilWaterParameterizations.md",
-        "Soil Heat Parameterizations" =>
-            "APIs/Land/SoilHeatParameterizations.md",
```
Comment (kpamnany) [inline docs/list_of_apis.jl:20]: Did you mean to remove this?

File: docs/list_of_apis.jl [hunk 10]
```diff
@@ -16,12 +16,9 @@ apis = [
...
     ],
     "Common" => [
         "Orientations" => "APIs/Common/Orientations.md",
-        "Spectra" => "APIs/Common/Spectra.md",
```
Comment (kpamnany) [inline docs/list_of_apis.jl:24]: And this?

**Comment by kpamnany** [comment]
Looks like you deleted the changes I made @akshaysridhar?

File: docs/list_of_apis.jl (hunk 10 above)
Comment (akshaysridhar) [reply docs/list_of_apis.jl:24]: No - I did not .. I think I messed up the rebase

**Review by kpamnany — APPROVED** [review]

**Pull request merged**